source/fuzz/transformation_propagate_instruction_down.h

```diff
@@ -0,0 +1,136 @@
+  //   GetInstructionToPropagate method).
+  // - the block has at least one acceptable successor (see
+  //   GetAcceptableSuccessors method).
+  // - none of the acceptable successors have OpPhi instructions that use the
+  //   original instruction.
+  // - it is possible to replace every use of the original instruction with some
+  //   of the propagated instructions (or an OpPhi if we can create it - see
+  //   CanAddOpPhiInstruction method).
+  static bool IsApplicableToBlock(opt::IRContext* ir_context,
+                                  uint32_t block_id);
+
+  // Returns ids of successors of |block_id|, that can be used to propagate an
+  // instruction into. Concretely, a successor block is acceptable if all
+  // dependencies of the propagated instruction dominate it. Note that this
+  // implies that an acceptable successor must be reachable in the CFG.
+  // For example:
+  //    %1 = OpLabel
+  //         OpSelectionMerge %2 None
+  //         OpBranchConditional %cond %2 %3
+  //    %3 = OpLabel
+  //    %4 = OpUndef %int
+  //    %5 = OpCopyObject %int %4
+  //         OpBranch %2
+  //    %2 = OpLabel
+  //    ...
+  // In this example, %2 is not an acceptable successor of %3 since one of the
+  // dependencies (%4) of the propagated instruction (%5) does not dominate it.
+  static std::unordered_set<uint32_t> GetAcceptableSuccessors(
+      opt::IRContext* ir_context, uint32_t block_id);
+
+  std::unordered_set<uint32_t> GetFreshIds() const override;
+
+ private:
+  // Returns the last possible instruction in the |block_id| that satisfies the
+  // following properties:
+  // - has result id
+  // - has type id
+  // - has supported opcode (see IsOpcodeSupported method)
+  // - has no users in its basic block.
+  // Returns nullptr if no such an instruction exists.
+  static opt::Instruction* GetInstructionToPropagate(opt::IRContext* ir_context,
+                                                     uint32_t block_id);
+
+  // Returns true if |opcode| is supported by this transformation.
+  static bool IsOpcodeSupported(SpvOp opcode);
+
+  // Returns the first instruction in the |block| that allows us to insert
+  // |opcode| above itself. Returns nullptr is no such instruction exists.
+  static opt::Instruction* GetFirstInsertBeforeInstruction(
+      opt::IRContext* ir_context, uint32_t block_id, SpvOp opcode);
+
+  // Returns true if we can add an OpPhi instruction that groups all the
+  // propagated clones of the original instruction. |maybe_header_block_id| is a
+  // result id of the block we propagate the instruction from. |successor_ids|
+  // contains result ids of the successors we propagate the instruction into.
+  // Concretely, returns true if:
+  // - |maybe_header_block_id| is a header block
+  // - the header's merge block is reachable
+  // - there must be at least one |maybe_header_block_id|'s acceptable successor
+  //   for every predecessor of the merge block, dominating that predecessor.
```

Comment:
A small example would really help here, showing where this OpPhi instruction would go.

Also, I think this could be more general (but not sure it's worth making it more general).

Suppose we had a loop of this form:

```
     
     |
   HEADER <---
     |        \
    BLOCK      |
    /   \      |
   A    B---CONTINUE
    \  /
    MERGE

then presumably if we propagated instructions from BLOCK to A and B we could Phi them together again at MERGE?

Reply:
Yes, I think it's worthwhile to make it more general. This would allow us, for example, to add OpPhis in certain cases when we propagate an instruction from a back-edge block into a header block and the merge block is a successor of a back-edge block. Although, I think it's a somewhat unrealistic example. Nevertheless, the change is not hard.

Fixed in b6fe77b.